[python-package] Build from Sources failed

## Description
Trying to build from sources under python-package but got following error message:

```text
(base) PS F:\src\exp\LightGBM\python-package> pip install --no-binary lightgbm lightgbm
Defaulting to user installation because normal site-packages is not writeable
DEPRECATION: --no-binary currently disables reading from the cache of locally built wheels. In the future --no-binary will not influence the wheel cache. pip 23.1 will enforce this behaviour change. A possible replacement is to use the --no-cache-dir option. You can use the flag --use-feature=no-binary-enable-wheel-cache to test the upcoming behaviour. Discussion can be found at https://github.com/pypa/pip/issues/11453
Collecting lightgbm
  Using cached lightgbm-3.3.5.tar.gz (1.5 MB)
  Preparing metadata (setup.py) ... done
Requirement already satisfied: wheel in c:\programdata\miniconda3\lib\site-packages (from lightgbm) (0.38.4)
Requirement already satisfied: numpy in c:\programdata\miniconda3\lib\site-packages (from lightgbm) (1.24.3)
Requirement already satisfied: scipy in c:\programdata\miniconda3\lib\site-packages (from lightgbm) (1.10.1)
Requirement already satisfied: scikit-learn!=0.22.0 in c:\programdata\miniconda3\lib\site-packages (from lightgbm) (1.2.2)
Requirement already satisfied: joblib>=1.1.1 in c:\programdata\miniconda3\lib\site-packages (from scikit-learn!=0.22.0->lightgbm) (1.2.0)
Requirement already satisfied: threadpoolctl>=2.0.0 in c:\programdata\miniconda3\lib\site-packages (from scikit-learn!=0.22.0->lightgbm) (3.1.0)
Installing collected packages: lightgbm
  DEPRECATION: lightgbm is being installed using the legacy 'setup.py install' method, because the '--no-binary' option was enabled for it and this currently disables local wheel building for projects that don't have a 'pyproject.toml' file. pip 23.1 will enforce this behaviour change. A possible replacement is to enable the '--use-pep517' option. Discussion can be found at https://github.com/pypa/pip/issues/11451
  Running setup.py install for lightgbm ... error
  error: subprocess-exited-with-error

  × Running setup.py install for lightgbm did not run successfully.
  │ exit code: 1
  ╰─> [32 lines of output]
      INFO:root:running install
      C:\ProgramData\miniconda3\lib\site-packages\setuptools\command\install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
        warnings.warn(
      INFO:LightGBM:Starting to compile the library.
      INFO:LightGBM:Starting to compile with MSBuild from existing solution file.
      WARNING:LightGBM:Compilation with MSBuild from existing solution file failed.
      INFO:LightGBM:Starting to compile with Visual Studio 16 2019 (x64).
      INFO:LightGBM:Starting to compile with Visual Studio 15 2017 (x64).
      INFO:LightGBM:Starting to compile with Visual Studio 14 2015 (x64).
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "C:\Users\ychiu\AppData\Local\Temp\pip-install-1g7ts6p9\lightgbm_3a2cce6a9b294f82832195b7b67412d8\setup.py", line 334, in <module>
          setup(name='lightgbm',
        File "C:\ProgramData\miniconda3\lib\site-packages\setuptools\__init__.py", line 87, in setup
          return distutils.core.setup(**attrs)
        File "C:\ProgramData\miniconda3\lib\site-packages\setuptools\_distutils\core.py", line 185, in setup
          return run_commands(dist)
        File "C:\ProgramData\miniconda3\lib\site-packages\setuptools\_distutils\core.py", line 201, in run_commands
          dist.run_commands()
        File "C:\ProgramData\miniconda3\lib\site-packages\setuptools\_distutils\dist.py", line 969, in run_commands
          self.run_command(cmd)
        File "C:\ProgramData\miniconda3\lib\site-packages\setuptools\dist.py", line 1208, in run_command
          super().run_command(command)
        File "C:\ProgramData\miniconda3\lib\site-packages\setuptools\_distutils\dist.py", line 988, in run_command
          cmd_obj.run()
        File "C:\Users\ychiu\AppData\Local\Temp\pip-install-1g7ts6p9\lightgbm_3a2cce6a9b294f82832195b7b67412d8\setup.py", line 248, in run
          compile_cpp(use_mingw=self.mingw, use_gpu=self.gpu, use_cuda=self.cuda, use_mpi=self.mpi,
        File "C:\Users\ychiu\AppData\Local\Temp\pip-install-1g7ts6p9\lightgbm_3a2cce6a9b294f82832195b7b67412d8\setup.py", line 192, in compile_cpp
          raise Exception("\n".join(('Please install Visual Studio or MS Build and all required dependencies first',
      Exception: Please install Visual Studio or MS Build and all required dependencies first
      The full version of error log was saved into C:\Users\ychiu\LightGBM_compilation.log
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> lightgbm

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.

I checked C:\Users\ychiu\LightGBM_compilation.log:

CMake Error at CMakeLists.txt:32 (PROJECT):
  Generator

    Visual Studio 16 2019

  could not find any instance of Visual Studio.

-- Configuring incomplete, errors occurred!
CMake Error at CMakeLists.txt:32 (PROJECT):
  Generator

    Visual Studio 15 2017

  could not find any instance of Visual Studio.

-- Configuring incomplete, errors occurred!
-- Selecting Windows SDK version  to target Windows 10.0.19045.
CMake Error at CMakeLists.txt:32 (PROJECT):
  Failed to run MSBuild command:

    MSBuild.exe

  to get the value of VCTargetsPath:

    The system cannot find the file specified

-- Configuring incomplete, errors occurred!
```

On my computer, I have Visual Studio 2022 & cmake 3.26.4 installed,
also I have made sure cmake is in system PATH that I was able to following instructions on windows build (not python-package) without problem.

## Reproducible example
<!-- Minimal code that exhibits this behavior -->

```shell
pip install --no-binary lightgbm lightgbm
```

## Environment info

LightGBM version or commit hash:

The latest one I got by using:

```shell
git clone --recursive https://github.com/microsoft/LightGBM
```

Command(s) you used to install LightGBM

```shell
git clone --recursive https://github.com/microsoft/LightGBM
```

<!-- Put any additional environment information here -->

I checked the registry: 
`HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\MSBuild\ToolsVersions\4.0`

originally there is no VCTargetsPath there and I manually added VCTargetsPath with value `$(MSBuildExtensionsPath32)\Microsoft.Cpp\v4.0\v140\`

but still got the same error message :(

## Additional Comments
<!-- What else should we know? -->

Thanks James for your response :)

1. 
I was trying to build the lightgbm python package from the source code which I cloned from GitHub.
Because I got error message when I ran 
 ```
pip install --no-binary lightgbm lightgbm
```
I tried to lightgbm windows build to verify if my cmake & visual studio 2022 have any issues.

2. Here is the output I got:
```
MINGW64 /f/src/exp/LightGBM/build (master)
$ cmake ..
\-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.19045.
-- The C compiler identification is MSVC 19.36.32532.0
-- The CXX compiler identification is MSVC 19.36.32532.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found OpenMP_C: -openmp (found version "2.0")
-- Found OpenMP_CXX: -openmp (found version "2.0")
-- Found OpenMP: TRUE (found version "2.0")
-- Performing Test MM_PREFETCH
-- Performing Test MM_PREFETCH - Failed
-- Performing Test MM_MALLOC
-- Performing Test MM_MALLOC - Failed
-- Looking for inet_pton
-- Looking for inet_pton - found
-- Configuring done (16.6s)
-- Generating done (0.1s)
-- Build files have been written to: F:/src/exp/LightGBM/build

 MINGW64 /f/src/exp/LightGBM/build (master)
$ make -j2
make: *** No targets specified and no makefile found.  Stop.
```
seems cmake succeeded but make failed :( 
it was good when I used:
```
git clone --recursive https://github.com/microsoft/LightGBM
cd LightGBM
mkdir build
cd build
cmake -A x64 -DUSE_OPENMP=OFF ..
cmake --build . --target ALL_BUILD --config Release
```
or
```
git clone --recursive https://github.com/microsoft/LightGBM
cd LightGBM
mkdir build
cd build
cmake -A x64 -DUSE_MPI=ON ..
cmake --build . --target ALL_BUILD --config Release
```

> make: *** No targets specified and no makefile found.  Stop.

You missed running `cd build` before `make -j2`.

> trying to build from sources in GitHub

Ok, please try running the following from the root of the repository.

```shell
sh build-python.sh install
```

Don't run `pip install lightgbm` if your goal is to build from source in this repo. That command downloads the source code from PyPI.

--- a/.ci/test_windows.ps1
+++ b/.ci/test_windows.ps1
@@ -101,9 +101,9 @@ elseif ($env:TASK -eq "bdist") {
 } elseif (($env:APPVEYOR -eq "true") -and ($env:TASK -eq "python")) {
   cd $env:BUILD_SOURCESDIRECTORY
   if ($env:COMPILER -eq "MINGW") {
-    sh $env:BUILD_SOURCESDIRECTORY/build-python.sh install --mingw ; Check-Output $?
+    sh $env:BUILD_SOURCESDIRECTORY/build-python.sh install --user --mingw ; Check-Output $?
   } else {
-    sh $env:BUILD_SOURCESDIRECTORY/build-python.sh install ; Check-Output $?
+    sh $env:BUILD_SOURCESDIRECTORY/build-python.sh install --user; Check-Output $?
   }
 }
 

--- a/build-python.sh
+++ b/build-python.sh
@@ -333,6 +333,7 @@ if test "${INSTALL}" = true; then
     cd ../dist
     pip install \
         ${PIP_INSTALL_ARGS} \
+        --force-reinstall \
         --find-links=. \
         lightgbm
     cd ../

